### ISSUE TITLE
Case where `WordPress.Security.ValidatedSanitizedInput.InputNotSanitized` might be a bit counter-intuitive

### ISSUE THREAD

**Issue by leonidasmi** [WordPress/WordPress-Coding-Standards#2246]
## Bug Description

The `WordPress.Security.ValidatedSanitizedInput.InputNotSanitized` error is thrown when a GET parameter is unslashed and then only used in a direct comparison, but it's not thrown when it's only used in a direct comparison.

I would expect the same behavior between those two cases, ideally and more specifically for that error to *not* appear in both cases.
## Minimal Code Snippet

The issue happens when running this command:
```bash
phpcs ...
```

... over a file containing this code:
```php
if ( isset( $_GET['foo'] ) && wp_unslash( $_GET['foo'] ) === 'bar' ) {
	//code;
}
```

but not over a file containing this code:

```php
if ( isset( $_GET['foo'] ) && $_GET['foo'] === 'bar' ) {
	//code;
}
```

## Error Code

```
ERROR | [ ] Detected usage of a non-sanitized input variable:
    |       |     $_GET['foo']
    |       |     (WordPress.Security.ValidatedSanitizedInput.InputNotSanitized)
```

It would probably make sense if the error should not appear in both cases? 🙂 

## Environment

| Question               | Answer
| ------------------------| -------
| PHP version             | 7.4.20
| PHP_CodeSniffer version | 3.7.2
| WPCS version            | 2.3.0
| WPCS install type       | Composer project local and git clone (haven't tested others)

## Tested Against `develop` branch?
- [ ] I have verified the issue still exists in the `develop` branch of WPCS.

**Comment by owaisahmed5300** [comment]
yes, also if this is POST/REQUEST/SERVER/COOKIE. not only get.

As long as this is a variable existence check. this should not raise Sanitize error.

**Comment by GaryJones** [comment]
Have you tried with the `develop` branch?

**Comment by jrfnl** [comment]
I know this is something which still needs looking at, I have some ideas, but nothing ready to pull yet.

Idea is along the lines of:
* Check if the use is within a control structure condition.
* If so, use a while loop to go though the nested parentheses from deepest to max the control structure parentheses
* For each parentheses, do a number of checks, like "in comparison" (for the token before the current open parens + the close parens)
* Check if the function being used is not one which changes the param by reference (in as far as possible, i.e. PHP native functions only + based on Reflection)

**Comment by owaisahmed5300** [comment]
@ve3 esc_url_raw() would be used to validate URLs before saving them into database.

**Comment by banago** [comment]
This is still happening in the **Plugin Check** plugin.

**Comment by jrfnl** [comment]
@kkmuffme Because your comment has absolutely nothing to do with the original issue and is just (unwelcome) noise.
While it may be about the same sniff, it is not addressing the same issue and the issue you are talking about is one which is unlikely to ever be addressed at all due to the complexities it brings - which you completely ignore and show understanding off.
